### ISSUE TITLE
Using set_fact with ansible_ssh_host no longer works in devel

### ISSUE THREAD

**Issue by dstanek** [ansible/ansible#11436]
I created a very [simple playbook](http://paste.openstack.org/show/326119/) to highlight the issues I am seeing. The most relevant part is:

``` yaml
    - name: Update host IP in inventory
      set_fact:
        ansible_ssh_host: "{{ instance.public_ip[0] }}"
      always_run: yes
```

In my playbook I create/ensure an OpenStack instance exists and change the IP address in the inventory so that I can actually talk to it. Then I try to run a few commands.

This works in 1.9.1, but not in devel. It looks like in the 1.9 branch the code is [explicitly looking for ansible_ssh_host](https://github.com/ansible/ansible/blob/stable-1.9/lib/ansible/runner/__init__.py#L877), but I can't find the equivalent code in devel.

**Comment by bcoca** [comment]
we rewrote the core engine in v2 (current devel) so some features have not been ported yet. This is a bug and it should be present and working.

**Comment by jimi-c** [comment]
Hi @dstanek, could you try this with the most recent pull of devel? We've recently made quite a few improvements/fixes here, as bcoca mentioned, so I'd like to know if this is still a problem.

Thanks!

**Comment by tkinz27** [comment]
I'm still seeing the issue.

```
TASK [os_setup : os_setup : add facts from nova] ********************************
ok: [os-test] => {"ansible_facts": {"ansible_ssh_host": "10.17.216.195", "ansible_ssh_private_key_file": "keys/my-os.pem"}, "changed": false}

TASK [os_setup : os_setup : debug var=ansible_ssh_host] *************************
ok: [os-test] => {
    "ansible_ssh_host": "os-test",
    "changed": false
}
```

```
$ ansible --version
ansible 2.0.0 (devel 048cd755d9) last updated 2015/07/21 20:10:54 (GMT +000)
  lib/ansible/modules/core: (detached HEAD 55bc9e8fb1) last updated 2015/07/21 20:16:05 (GMT +000)
  lib/ansible/modules/extras: (detached HEAD 26c1bd76c3) last updated 2015/07/21 20:16:05 (GMT +000)
  v1/ansible/modules/core:  not found - use git submodule update --init v1/ansible/modules/core
  v1/ansible/modules/extras:  not found - use git submodule update --init v1/ansible/modules/extras
  configured module search path = /usr/share/ansible:./modules
akinsley - ~/hydra/ansible
```

**Comment by tkinz27** [comment]
Some more info... here is the relevant playbook

```
- name: add facts from nova
  local_action:
    module: set_fact
    ansible_ssh_host: "{{ cloud.server.addresses[os_public_network][0].addr }}"
    ansible_ssh_private_key_file: "{{ os_private_key_file }}"
    stuff: testing

- debug: var=ansible_ssh_host
- debug: var=ansible_ssh_private_key_file
- debug: var=stuff
```

None of the variables are getting set. That did work not too long ago.

**Comment by bcoca** [comment]
you need to indent them more

**Comment by jimi-c** [comment]
@tkinz27 that's odd, running your example I see things set properly:

<pre>
TASK [add facts from nova] ******************************************************
ok: [localhost] => {"ansible_facts": {"ansible_ssh_host": "1.1.1.1", "ansible_ssh_private_key_file": "/path/to/key/file", "stuff": "testing"}, "changed": false}
TASK [debug var=ansible_ssh_host] ***********************************************
ok: [localhost] => {
    "ansible_ssh_host": "1.1.1.1", 
    "changed": false
}
TASK [debug var=ansible_ssh_private_key_file] ***********************************
ok: [localhost] => {
    "ansible_ssh_private_key_file": "/path/to/key/file", 
    "changed": false
}
TASK [debug var=stuff] **********************************************************
ok: [localhost] => {
    "changed": false, 
    "stuff": "testing"
}
</pre>

**Comment by tkinz27** [comment]
Could it be that I am running the set_fact within a role? I describe my setup more here https://groups.google.com/forum/#!topic/ansible-devel/a4oGoLteON0

Before the set_fact would work within the role and then the variable would revert back after the role completed. Today the set_fact failed all together though.

**Comment by jimi-c** [comment]
I would not expect that to make a difference.

**Comment by tkinz27** [comment]
OK interesting so I changed my role tasks/main.yml to remove the local_action call from the set_fact and the facts are now getting set correctly.

But I'm back to my original problem described in the google groups above. Basically once I move outside of the role, the ansible_ssh_host reverts back to the original value.

```
TASK [os_setup : add facts from nova] *******************************************
ok: [os-test] => {"ansible_facts": {"ansible_ssh_host": "10.17.216.195", "ansible_ssh_private_key_file": "keys/ccs-os.pem", "stuff": "testing"}, "changed": false}

TASK [os_setup : debug var=ansible_ssh_host] ************************************
ok: [os-test] => {
    "ansible_ssh_host": "10.17.216.195",
    "changed": false
}

TASK [os_setup : debug var=ansible_ssh_private_key_file] ************************
ok: [os-test] => {
    "ansible_ssh_private_key_file": "keys/ccs-os.pem",
    "changed": false
}

TASK [os_setup : debug var=stuff] ***********************************************
ok: [os-test] => {
    "changed": false,
    "stuff": "testing"
}

TASK [os_setup : wait for ssh] **************************************************
ESTABLISH LOCAL CONNECTION FOR USER: sgadmin
127.0.0.1 EXEC mkdir -p "$HOME/.ansible/tmp/ansible-tmp-1437602902.03-252345998855787" && chmod a+rx "$HOME/.ansible/tmp/ansible-tmp-1437602902.03-252345998855787" && echo "$HOME/.ansible/tmp/ansible-tmp-1437602902.03-252345998855787"
127.0.0.1 PUT /tmp/tmpVjA0hG TO /cip/.ansible/tmp/ansible-tmp-1437602902.03-252345998855787/wait_for
127.0.0.1 EXEC LANG=C LC_MESSAGES=C LC_CTYPE=C /usr/bin/python /cip/.ansible/tmp/ansible-tmp-1437602902.03-252345998855787/wait_for; rm -rf "/cip/.ansible/tmp/ansible-tmp-1437602902.03-252345998855787/" >/dev/null 2>&1
ok: [os-test] => {"changed": false, "elapsed": 0, "path": null, "port": 22, "search_regex": null, "state": "started"}

PLAY ****************************************************************************

TASK [setup] ********************************************************************
<os-test> ESTABLISH SSH CONNECTION FOR USER: centos
<os-test> EXEC ssh -C -tt -vvv -o ControlMaster=auto -o ControlPersist=60s -o ControlPath="/tmp/ansible-ssh-%h-%p-%r" -o StrictHostKeyChecking=no -o Port=22 -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=centos -o ConnectTimeout=10 os-test mkdir -p "$HOME/.ansible/tmp/ansible-tmp-1437602902.53-67574716631867" && chmod a+rx "$HOME/.ansible/tmp/ansible-tmp-1437602902.53-67574716631867" && echo "$HOME/.ansible/tmp/ansible-tmp-1437602902.53-67574716631867"
fatal: [os-test]: FAILED! => {
    "failed": true,
    "msg": "ERROR! SSH encountered an unknown error. The output was:\nOpenSSH_6.6.1, OpenSSL 1.0.1f 6 Jan 2014\r\ndebug1: Reading configuration data /etc/ssh/ssh_config\r\ndebug1: /etc/ssh/ssh_config line 19: Applying options for *\r\ndebug1: auto-mux: Trying existing master\r\ndebug1: Control socket \"/tmp/ansible-ssh-os-test-22-centos\" does not exist\r\nssh: Could not resolve hostname os-test: Name or service not known\r\n"
}
```

Just for clarity, my new playbook shows this

```
- name: add facts from nova
  set_fact:
    ansible_ssh_host: "{{ cloud.server.addresses[os_public_network][0].addr }}"
    ansible_ssh_private_key_file: "{{ os_private_key_file }}"
    stuff: testing
```

instead of the syntax I used above.

**Comment by jimi-c** [comment]
@tkinz27 ahh yes, `local_action` is just shorthand for `delegate_to: localhost`, so you were essentially setting facts on localhost instead of the inventory host. I'll see if I can reproduce the issue with roles myself, but that seems very odd. My guess is that we're just not getting all of the vars when setting up the connection information, as they are currently only coming from the host and task objects. The vars setup by set_fact go in their own pool of variables (known as the vars_cache, in the VariableManager), which probably isn't being brought in when we're looking for those special variables.

**Comment by jimi-c** [comment]
@tkinz27 still unable to reproduce this. Just to be clear, I know I asked @dstanek to try the devel version, but are you also running the latest commit from devel?

**Comment by amenonsen** [comment]
I also can't reproduce using latest devel(@a1a8997e) with the play listed earlier in this ticket.

**Comment by andreobrown** [comment]
TLDR;
Setting ansible_ssh_host in play does not take. SSH connections are still issues to the DNS name (and DNS resolved IP address).

Setting ansible_ssh_host using set_fact does not take. The subsequent tasks connect using the hostname from the inventory file, even though a debug shows the correct value being set. 

roles/networking/tasks/main.yml

```
- name: set ssh ip address
  set_fact:
    ansible_ssh_host: "{{ staging_ip }}"

- debug: msg="SSH address {{ansible_ssh_host}}"
```

Output from the play:

```
ok: [a-testing.redacted.com] => {"changed": false, "elapsed": 10, "path": null, "port": 22, "search_regex": "OpenSSH", "state": "started"}
ok: [b-testing.redacted.com] => {"changed": false, "elapsed": 10, "path": null, "port": 22, "search_regex": "OpenSSH", "state": "started"}
ok: [c-testing.redacted.com] => {"changed": false, "elapsed": 10, "path": null, "port": 22, "search_regex": "OpenSSH", "state": "started"}
ok: [d-testing.redacted.com] => {"changed": false, "elapsed": 10, "path": null, "port": 22, "search_regex": "OpenSSH", "state": "started"}

TASK [networking : set ssh ip address] ******************************************
ok: [b-testing.redacted.com] => {"ansible_facts": {"ansible_ssh_host": "192.30.17.252"}, "changed": false}
ok: [c-testing.redacted.com] => {"ansible_facts": {"ansible_ssh_host": "192.30.17.253"}, "changed": false}
ok: [a-testing.redacted.com] => {"ansible_facts": {"ansible_ssh_host": "192.30.17.251"}, "changed": false}
ok: [d-testing.redacted.com] => {"ansible_facts": {"ansible_ssh_host": "192.30.17.254"}, "changed": false}

TASK [networking : debug msg=SSH address {{ansible_ssh_host}}] ******************
ok: [a-testing.redacted.com] => {
    "changed": false,
    "msg": "SSH address 192.30.17.251"
}
ok: [b-testing.redacted.com] => {
    "changed": false,
    "msg": "SSH address 192.30.17.252"
}
ok: [c-testing.redacted.com] => {
    "changed": false,
    "msg": "SSH address 192.30.17.253"
}
ok: [d-testing.redacted.com] => {
    "changed": false,
    "msg": "SSH address 192.30.17.254"
}
```

Debug from the failed SSH connection:

```
fatal: [a-testing.redacted.com]: FAILED! => {
    "failed": true,
    "msg": "ERROR! SSH encountered an unknown error. The output was:\nOpenSSH_6.2p2, OSSLShim 0.9.8r 8 Dec 2011\ndebug1: Reading configuration data /etc/ssh_config\r\ndebug1: /etc/ssh_config line 20: Applying options for *\r\ndebug1: /etc/ssh_config line 102: Applying options for *\r\ndebug1: auto-mux: Trying existing master\r\ndebug1: Control socket \"/tmp/ansible-ssh-a-testing.redacted.com-22-redacted\" does not exist\r\ndebug2: ssh_connect: needpriv 0\r\ndebug1: Connecting to a-testing.redacted.com [207.xxx.xx.xx] port 22.\r\ndebug2: fd 3 setting O_NONBLOCK\r\ndebug1: connect to address 207.xxx.xx.xx port 22: Operation timed out\r\nssh: connect to host a-testing.redacted.com port 22: Operation timed out\r\n"
}
```

Perhaps related is that passing ansible_ssh_host as a variable to an include also does not work. No matter how I set ansible_ssh_host in a play, it seems to ignore it and use the dns name.

In my site.yml

```
# Configure networking
- include: configure-networking-new-vms.yml tags=newnetworking ansible_ssh_host="{{ staging_ip }}"
```

Output from the run:

```
TASK [networking : debug msg=SSH address {{ansible_ssh_host}}] ******************
ok: [a-testing.redacted.com] => {
    "changed": false,
    "msg": "SSH address 192.30.17.251"
}
ok: [c-testing.redacted.com] => {
    "changed": false,
    "msg": "SSH address 192.30.17.254"
}
ok: [d-testing.redacted.com] => {
    "changed": false,
    "msg": "SSH address 192.30.17.252"
}
ok: [b-testing.redacted.com] => {
    "changed": false,
    "msg": "SSH address 192.30.17.253"
}
```

Debug output:

```
fatal: [a-testing.redacted.com]: FAILED! => {
    "failed": true,
    "msg": "ERROR! SSH encountered an unknown error. The output was:\nOpenSSH_6.2p2, OSSLShim 0.9.8r 8 Dec 2011\ndebug1: Reading configuration data /etc/ssh_config\r\ndebug1: /etc/ssh_config line 20: Applying options for *\r\ndebug1: /etc/ssh_config line 102: Applying options for *\r\ndebug1: auto-mux: Trying existing master\r\ndebug1: Control socket \"/tmp/ansible-ssh-a-testing.redacted.com-22-redacted\" does not exist\r\ndebug2: ssh_connect: needpriv 0\r\ndebug1: Connecting to a-testing.redacted.com [207.xxx.xx.xx] port 22.\r\ndebug2: fd 3 setting O_NONBLOCK\r\ndebug1: connect to address 207.xxx.xx.xx port 22: Operation timed out\r\nssh: connect to host a-testing.redacted.com port 22: Operation timed out\r\n"
}
```

Setting ansible_ssh_host in the inventory file appears to work, but is not acceptable for my use case as the IP address of the host changes in the play (it's initially set to a known value, then changed dynamically to the correct value).

Here is my ansible version:

```
ansible-playbook 2.0.0 (devel a1a8997e89) last updated 2015/07/27 01:07:25 (GMT -500)
  lib/ansible/modules/core: (detached HEAD 2bfddb015d) last updated 2015/07/27 01:07:36 (GMT -500)
  lib/ansible/modules/extras: (detached HEAD 14fffa5b53) last updated 2015/07/25 15:24:26 (GMT -500)
  v1/ansible/modules/core: (detached HEAD f8d8af17cd) last updated 2015/07/25 15:24:32 (GMT -500)
  v1/ansible/modules/extras: (detached HEAD 495ad450e5) last updated 2015/07/25 15:24:37 (GMT -500)
  config file = /Users/andbro/ansible.cfg
  configured module search path = None
```

**Comment by dstanek** [comment]
This issue still exists for me in devel. I think I'm seeing the same thing as @andreobrown.

Ansible is trying to ssh using the hostname instead of using the IP address acquired from Nova. In the output below you can see that it is trying to use the hostname instead of IP.

```
PLAY [check instance] ***********************************************************

TASK [setup] ********************************************************************
<dstanek-tempest> ESTABLISH SSH CONNECTION FOR USER: root
<dstanek-tempest> EXEC ssh -C -vvv -o ControlMaster=auto -o ControlPersist=60s -o ControlPath="/tmp/ansible-ssh-%h-%p-%r" -o StrictHostKeyChecking=no -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=root -o ConnectTimeout=10 dstanek-tempest LANG=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 LC_CTYPE=en_US.UTF-8 /usr/bin/python
fatal: [dstanek-tempest]: FAILED! => {"failed": true, "msg": "ERROR! SSH Error: data could not be sent to the remote host. Make sure this host can be reached over ssh"}
```

Version I am testing with:

```
> ansible-playbook --version
 ansible-playbook 2.0.0 (devel 9416fc6271) last updated 2015/07/27 15:51:33 (GMT -400)
  lib/ansible/modules/core: (detached HEAD bc60c893f5) last updated 2015/07/27 15:51:45 (GMT -400)
  lib/ansible/modules/extras: (detached HEAD 14fffa5b53) last updated 2015/07/27 15:51:45 (GMT -400)
  v1/ansible/modules/core: (detached HEAD f8d8af17cd) last updated 2015/06/10 07:49:51 (GMT -400)
  v1/ansible/modules/extras: (detached HEAD 495ad450e5) last updated 2015/06/10 07:49:59 (GMT -400)
  config file = /Users/dstanek/.ansible.cfg
  configured module search path = ~/Projects/ansible/library
```

**Comment by tkinz27** [comment]
@jimi-c The last version I last tested with was

```
$ ansible --version
ansible 2.0.0 (devel d412bc72ef) last updated 2015/07/22 22:04:14 (GMT +000)
  lib/ansible/modules/core: (detached HEAD 222927256d) last updated 2015/07/22 22:04:19 (GMT +000)
  lib/ansible/modules/extras: (detached HEAD 167cad99fa) last updated 2015/07/22 22:04:19 (GMT +000)
```

Today is not a good day for me to pull the latest and try again, but hopefully I can try again tomorrow.

**Comment by dstanek** [comment]
I'm having a hard time following the code, so I may be way off....

It looks like there is [functionality](https://github.com/ansible/ansible/blob/stable-1.9/lib/ansible/runner/__init__.py#L844) in the stable-1.9 runner that explicitly looks for the ansible_ssh_\* variables and properly sets them for the connection. I can't find this logic in the devel branch.

From what I can tell the set_fact works fine. The issue is that it's never actually used.

**Comment by tkinz27** [comment]
So just verified that it does not work with the latest devel.

```
ansible 2.0.0 (devel 3b9282fb90) last updated 2015/07/28 18:34:34 (GMT +000)
  lib/ansible/modules/core: (detached HEAD a46b3a4dc4) last updated 2015/07/28 18:34:41 (GMT +000)
  lib/ansible/modules/extras: (detached HEAD e30d8b84fe) last updated 2015/07/28 18:34:41 (GMT +000)
  v1/ansible/modules/core:  not found - use git submodule update --init v1/ansible/modules/core
  v1/ansible/modules/extras:  not found - use git submodule update --init v1/ansible/modules/extras
  config file =
  configured module search path = None
```

I moved my role into a single playbook and was able to reproduce the issue. 

```
---

- hosts: all
  gather_facts: no
  vars:
    os_public_network: my-net
    os_flavor: m1.tiny.m1g.d10g
    os_image: CentOS-7-x86_64-gold-v1
    os_key_name: ccs-os
    os_nics: [ net-name: my-net ]
    os_security_groups: default
    os_private_key_file: keys/my-os.pem
    os_auto_floating_ip: no

  tasks:
    - name: create node in nova
      register: cloud
      local_action:
        module: os_server
        name: "{{ inventory_hostname }}"
        state: present
        image: "{{ os_image }}"
        key_name: "{{ os_key_name }}"
        flavor: "{{ os_flavor }}"
        nics: "{{ os_nics }}"
        security_groups: "{{ os_security_groups }}"
        auto_floating_ip: "{{ os_auto_floating_ip }}"


    - debug: var=cloud

    - name: add facts from nova
      set_fact:
        ansible_ssh_host: "{{ cloud.server.addresses[os_public_network][0].addr }}"
        ansible_ssh_private_key_file: "{{ os_private_key_file }}"
        stuff: testing

    - name: try to touch a file in first play
      file: state=touch path=/home/centos/first_play

- hosts: all
  user: centos

  tasks:

    - name: try to touch a file second play
      file: state=touch path=/home/centos/second_play
```

My hosts files is

```
$ cat ostest
[fun]
os-test
```

And the output is 

```
$ ansible-playbook -i ostest os_test.yml -vvvv
Using  as config file
2 plays in os_test.yml
Loaded callback default of type stdout, v2.0

PLAY ****************************************************************************

TASK [create node in nova] ******************************************************
ESTABLISH LOCAL CONNECTION FOR USER: root
localhost EXEC mkdir -p "$HOME/.ansible/tmp/ansible-tmp-1438109236.74-20325707842480" && echo "$HOME/.ansible/tmp/ansible-tmp-1438109236.74-20325707842480"
localhost PUT /tmp/tmp3t8k9m TO /cip/.ansible/tmp/ansible-tmp-1438109236.74-20325707842480/os_server
localhost EXEC LANG=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 LC_CTYPE=en_US.UTF-8 /usr/bin/python /cip/.ansible/tmp/ansible-tmp-1438109236.74-20325707842480/os_server; rm -rf "/cip/.ansible/tmp/ansible-tmp-1438109236.74-20325707842480/" >/dev/null 2>&1
ok: [os-test] => {"changed": false, "id": "0dd751dc-40cd-4445-aff5-57090a219130", "openstack": <redacted output>: []}}

TASK [debug var=cloud] **********************************************************
ok: [os-test] => {
    "changed": false,
    "cloud": {
        "changed": false,
......
<redacted unnecessary data>
......
    }
}

TASK [add facts from nova] ******************************************************
ok: [os-test] => {"ansible_facts": {"ansible_ssh_host": "10.17.216.195", "ansible_ssh_private_key_file": "keys/my-os.pem", "stuff": "testing"}, "changed": false}

TASK [try to touch a file in first play] ****************************************
<os-test> ESTABLISH SSH CONNECTION FOR USER: root
<os-test> EXEC ssh -C -tt -vvv -o ControlMaster=auto -o ControlPersist=60s -o ControlPath="/tmp/ansible-ssh-%h-%p-%r" -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o ConnectTimeout=10 os-test mkdir -p "$HOME/.ansible/tmp/ansible-tmp-1438109242.13-201449118629968" && echo "$HOME/.ansible/tmp/ansible-tmp-1438109242.13-201449118629968"
EXEC previous known host file not found for os-test
EXEC previous known host file not found for os-test
fatal: [os-test]: FAILED! => {"failed": true, "msg": "ERROR! SSH encountered an unknown error. The output was:\nOpenSSH_6.6.1, OpenSSL 1.0.1f 6 Jan 2014\r\ndebug1: Reading configuration data /etc/ssh/ssh_config\r\ndebug1: /etc/ssh/ssh_config line 19: Applying options for *\r\ndebug1: auto-mux: Trying existing master\r\ndebug1: Control socket \"/tmp/ansible-ssh-os-test-22-root\" does not exist\r\nssh: Could not resolve hostname os-test: Name or service not known\r\n"}
```

Hopefully this helps you guys reproduce the issue.

**Comment by jimi-c** [comment]
@dstanek / @tkinz27 Here is a patch which may resolve this for you. I have not tested this well, and is mainly based on analysis of the code: https://gist.github.com/jimi-c/38e049a5e17fbcd9b65e

**Comment by dstanek** [comment]
@jimi-c that's it! when i applied the patch i got the expected behavior. thanks!

**Comment by jimi-c** [comment]
# Closing This Ticket

Hi!

We believe the above commit should resolve this problem for you. This will also be included in the next major release.

If you continue seeing any problems related to this issue, or if you have any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:
- https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
- https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular issue is resolved.

Thank you!
